src/rgw/rgw_bl.cc

```diff
@@ -0,0 +1,984 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+/*
+ * Ceph - scalable distributed file system
+ *
+ * Copyright (C) 2017 Red Hat, Inc.
+ *
+ * This is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License version 2.1, as published by the Free Software
+ * Foundation.  See file COPYING.
+ *
+ */
+
+#include <string.h>
+#include <iostream>
+#include <map>
+
+#include <boost/algorithm/string/split.hpp>
+#include <boost/algorithm/string.hpp>
+
+#include <common/errno.h>
+#include "auth/Crypto.h"
+#include "cls/rgw/cls_rgw_client.h"
+#include "cls/lock/cls_lock_client.h"
+#include "rgw_common.h"
+#include "rgw_bucket.h"
+#include "rgw_bl.h"
+#include "rgw_rados.h"
+#include "common/Clock.h"
+#include "rgw_rest_client.h"
+#include "rgw_user.h"
+#include "rgw_acl_s3.h"
+#include "common/url_escape.h"
+
+
+#define dout_context g_ceph_context
+#define dout_subsys ceph_subsys_rgw
+
+using namespace std;
+using namespace librados;
+
+const char* BL_STATUS[] = {
+  "UNINITIAL",
+  "PROCESSING",
+  "FAILED",
+  "PERM_ERROR",
+  "ACL_ERROR",
+  "COMPLETE"
+};
+
+std::map<const string, const uint32_t> rgw_perm_map = boost::assign::map_list_of
+      ("READ", RGW_PERM_READ)
+      ("WRITE", RGW_PERM_WRITE)
+      ("READ_ACP", RGW_PERM_READ_ACP)
+      ("WRITE_ACP", RGW_PERM_WRITE_ACP)
+      ("FULL_CONTROL", RGW_PERM_FULL_CONTROL);
+
+void *RGWBL::BLWorker::entry() {
+  utime_t end = ceph_clock_now();
+  do {
+    utime_t start = ceph_clock_now();
+    if (should_work(start) &&
+        (start.sec() - end.sec() >= bl->deliver_interval)) {
+      dout(5) << "bucket logging deliver: start" << dendl;
+      int r = bl->process();
+      if (r < 0) {
+        dout(0) << "ERROR: bucket logging process() err=" << r << dendl;
+      }
+      dout(5) << "bucket logging deliver: stop" << dendl;
+    }
+    if (bl->going_down())
+      break;
+
+    end = ceph_clock_now();
+    int secs = schedule_next_start_time(end);
+    time_t next_time = end + secs;
+    char buf[30];
+    char *nt = ctime_r(&next_time, buf);
```

Comment:
there's `utime_t::asctime()` that can be used

Reply:
ok